tests-integration/lists_test.go

```diff
@@ -191,3 +191,167 @@ func TestUnreadCountMisordering(t *testing.T) {
+	// The following sequence of events take place:
+	// (1) r1: topic set
+	// (2) r1: message
+	// (3) r2: message
+	// (4) r3: message
+	// (5) r2: topic
+	// (6) r1: profile change
+
+	ts = ts.Add(time.Second)
+	r1Timeline = append(r1Timeline,
+		testutils.NewStateEvent(
+			t, "m.room.topic", "", alice, map[string]interface{}{"topic": "potato"}, testutils.WithTimestamp(ts),
+		),
+	)
+
+	ts = ts.Add(time.Second)
+	r1Timeline = append(r1Timeline,
+		testutils.NewMessageEvent(t, alice, "message in room 1", testutils.WithTimestamp(ts)),
+	)
+
+	ts = ts.Add(time.Second)
+	r2Timeline = append(r2Timeline,
+		testutils.NewMessageEvent(t, alice, "message in room 2", testutils.WithTimestamp(ts)),
+	)
+
+	ts = ts.Add(time.Second)
+	r3Timeline = append(r3Timeline,
+		testutils.NewMessageEvent(t, alice, "message in room 3", testutils.WithTimestamp(ts)),
+	)
+
+	ts = ts.Add(time.Second)
+	r2Timeline = append(r2Timeline,
+		testutils.NewStateEvent(
+			t, "m.room.topic", "", alice, map[string]interface{}{"topic": "bananas"}, testutils.WithTimestamp(ts),
+		),
+	)
+
+	ts = ts.Add(time.Second)
+	r1Timeline = append(r1Timeline,
+		testutils.NewStateEvent(
+			t, "m.room.member", alice, alice, map[string]interface{}{"membership": "join", "displayname": "all ice"}, testutils.WithTimestamp(ts),
+		),
+	)
+
+	const room1ID = "!room1:localhost"
+	r1 := roomEvents{
+		roomID: room1ID,
+		name:   "room 1",
+		state:  r1CreateState,
+		events: r1Timeline,
+	}
+	const room2ID = "!room2:localhost"
+	r2 := roomEvents{
+		roomID: room2ID,
+		name:   "room 2",
+		state:  r2CreateState,
+		events: r2Timeline,
+	}
+	const room3ID = "!room3:localhost"
+	r3 := roomEvents{
+		roomID: room3ID,
+		name:   "room 3",
+		state:  r3CreateState,
+		events: r3Timeline,
+	}
+
+	pqString := testutils.PrepareDBConnectionString()
+	v2 := runTestV2Server(t)
+	v3 := runTestServer(t, v2, pqString)
+	defer v2.close()
+	defer v3.close()
+
+	t.Log("We prepare to tell the proxy about these events.")
+	v2.addAccount(t, alice, aliceToken)
+	v2.queueResponse(aliceToken, sync2.SyncResponse{
+		Rooms: sync2.SyncRoomsResponse{
+			Join: v2JoinTimeline(r1, r2, r3),
+		},
+	})
+
+	t.Log("Alice requests a new sliding sync connection.")
+	v3.mustDoV3Request(t, aliceToken, sync3.Request{
+		Lists: map[string]sync3.RequestList{
+			"a": {
+				RoomSubscription: sync3.RoomSubscription{
+					TimelineLimit: 20,
+				},
+				Ranges: sync3.SliceRanges{{0, 2}},
+			},
+		},
+	})
+
+	// Confirm that the poller was polled from.
+	v2.waitUntilEmpty(t, aliceToken)
+
+	t.Log("The proxy restarts.")
+	v3.restart(t, v2, pqString)
+
+	// Vary the bump event types, and compare the room order we get to what we expect
+	cases := []struct {
+		BumpEventTypes []string
+		RoomIDs        []string
+	}{
+		{
+			BumpEventTypes: []string{"m.room.message"},
+			RoomIDs:        []string{room3ID, room2ID, room1ID},
+		},
+		{
+			BumpEventTypes: []string{"m.room.topic"},
+			RoomIDs:        []string{room2ID, room1ID, room3ID},
```

Comment:
Would be nice to xref values in the block comment above to justify the values you give here. I've checked them all but it would help if you could help the reader more. i.e:
```go
BumpEventTypes: []string{"m.room.topic"},
// r2 topic (5) then r1 topic (1) then no topic
RoomIDs:        []string{room2ID, room1ID, room3ID},
```